File: front/src/services/wallet/WalletService.ts
import type {
  PaymentMethod,
  PlanSubscription,
  SubscriptionMember,
  SubscriptionPlan,
  UserMonthlyAllowance,
  Wallet,
  WalletTransaction,
} from './types.ts'

export interface WalletService {
  myWallet(): Promise<Wallet | null>
  subscriptionPlans(): Promise<SubscriptionPlan[]>
  mySubscription(): Promise<PlanSubscription | null>
  myMonthlyAllowance(): Promise<UserMonthlyAllowance | null>
  initiateWalletTopUp(amountMinor: number, currency: string): Promise<WalletTransaction>
  subscribeToPlan(planId: string): Promise<PlanSubscription>
  cancelSubscription(id: string): Promise<PlanSubscription>
  addSubscriptionMember(
    subscriptionId: string,
    userId: string,
    isMinor: boolean,
  ): Promise<SubscriptionMember>
  removeSubscriptionMember(subscriptionId: string, userId: string): Promise<SubscriptionMember>
  setAutoRenew(subscriptionId: string, autoRenew: boolean): Promise<PlanSubscription>
  myPaymentMethods(): Promise<PaymentMethod[]>
  createSetupIntent(): Promise<string>
  addCard(paymentMethodId: string): Promise<PaymentMethod>
  removeCard(id: string): Promise<boolean>
}

File: back/src/main/kotlin/com/jvmvstv_v/back/wallet/service/WalletService.kt
package com.jvmvstv_v.back.wallet.service

import com.jvmvstv_v.back.wallet.model.PaymentMethod
import com.jvmvstv_v.back.wallet.model.PlanSubscription
import com.jvmvstv_v.back.wallet.model.SubscriptionMember
import com.jvmvstv_v.back.wallet.model.SubscriptionPlan
import com.jvmvstv_v.back.wallet.model.UserMonthlyAllowance
import com.jvmvstv_v.back.wallet.model.Wallet
import com.jvmvstv_v.back.wallet.model.WalletTransaction
import java.util.UUID

interface WalletService {
    fun myWallet(): Wallet?
    fun plans(): List<SubscriptionPlan>
    fun mySubscription(): PlanSubscription?
    fun myMonthlyAllowance(): UserMonthlyAllowance?
    fun initiateTopUp(amountMinor: Int, currency: String): WalletTransaction
    fun subscribeToPlan(planId: Int): PlanSubscription
    fun cancelSubscription(id: UUID): PlanSubscription
    fun addSubscriptionMember(subscriptionId: UUID, userId: UUID, isMinor: Boolean): SubscriptionMember
    fun removeSubscriptionMember(subscriptionId: UUID, userId: UUID): SubscriptionMember
    fun setAutoRenew(subscriptionId: UUID, autoRenew: Boolean): PlanSubscription
    fun myPaymentMethods(): List<PaymentMethod>
    fun createSetupIntent(): String
    fun addCard(paymentMethodId: String): PaymentMethod
    fun removeCard(id: UUID): Boolean
}
